modules/billing/payment-required/data.js

```diff
@@ -0,0 +1,103 @@
+import { useEffect } from 'react'
+import { autorun, observable } from 'mobx'
+import { useObserver } from 'mobx-react-lite'
+
+import { requestPremium as letter } from 'texts/letters'
+
+const globalContext = observable({
+  user: {},
+  organisation: {},
+})
+
+const userInput = observable({
+  update(patch) {
+    Object.assign(this, patch)
+  },
+})
+
+const emailContext = observable({
+  toEmail: 'enterp\u0072is\u0065\u0040cor\u0065\u002e\u0061\u0063\u002euk',
+
+  // These 4 are mirrored from the global store
+  // with and React's effect and MobX's autorun in useSync()
+  organisationId: '',
+  organisationName: '',
+
+  get fromName() {
+    return userInput.name ?? globalContext.user?.name
+  },
+
+  get fromEmail() {
+    return userInput.email ?? globalContext.user?.email
+  },
+
+  get context() {
+    return {
+      userName: this.fromName,
+      userEmail: this.fromEmail,
+      organisationName: globalContext.organisation.name,
+      organisationId: globalContext.organisation.id,
+    }
+  },
+
+  get subject() {
+    return userInput.subject ?? letter.subject.render(this.context)
+  },
+
+  get body() {
+    return userInput.body ?? letter.body.render(this.context)
+  },
+
+  get mailtoUrl() {
+    const { email, subject, body } = this
+    const params = new URLSearchParams({
+      subject,
+      body,
+    })
+    return `m\u0061il\u0074\u006f\u003a${email}?${params}`
+  },
+
+  update(...args) {
+    userInput.update(...args)
+  },
+})
+
+const useSync = (store) =>
+  useEffect(
+    () =>
+      autorun(() => {
```

Comment:
I see now why you wanted to export store. You made it real reactive. 👍 To me both store or this autorun is good. Maybe I would just correct my statement and say that exporting store is better because it doesn't make duplicate values even they are kept in sync. (Even if testing or whatsoever may be harder).

I am quite surprised why you decided to go this way.  Not saying it's a bad decision. I imagined using only `withGlobalStore` HOC and nothing else in modules. Importing mobx stuff here is making more files dependent on it.

Reply:
I wend this way because if the `user`, for example, is set to `null`, then to an object, then the object, we will not track it. Although, if we use `autorun` we will do.

This option looked kind of interesting to me, kind of fun. It does not have to be good, I do not really know what is better.

---

> I imagined using only `withGlobalStore` HOC and nothing else in modules. Importing mobx stuff here is making more files dependent on it.

I wanted to have some kind of shared local state. We repeat the form in a few places, so I wanted to have changes populated everywhere if the user edited a single one.

It saw 2 options for this: React Context and MobX. I went to MobX because it seemed to be easier, `useObserver` hook would automatically run updates for me. Although, uncontrolled components slightly prevent this idea. However, if use controlled inputs, we will need a more sophisticated solution.

Reply:
Using MobX was a good way to go. I would personally create a new sub store for it not mixing state and UI component together. You taught me that. 🙂